README.md
# Condor

[Terraform Module Registry Docs](https://registry.terraform.io/modules/vultr/condor/vultr/latest)

## Table of Contents
* [Requirements](#requirements)
* [Quick Start](#quick-start)
* [Firewall Configuration](#firewall-configuration)
  * [Control Plane HA VLB Firewall](#control-plane-ha-vlb-firewall)
  * [Cluster Nodes Vultr Firewall](#cluster-nodes-vultr-firewall)
* [Extensions](#extensions)
  * [Helm](#helm)
  * [Manifest Bundles](#manifest-bundles)
* [Limitagions](#limitagions)

## Requirements
  * A funded Vultr account and API Key, should be configured as an environment variable to be consumed by the Terraform provider:
  ```sh
  export VULTR_API_KEY="<api-key-here>"
  ```
  * The [k0sctl](https://github.com/k0sproject/k0sctl) executable must be installed and in your executable path.
  * A configured ssh-agent, i.e.:
  ```sh
  ssh-add ~/.ssh/id_rsa
  ```

## Cloud Addons
  * Installs the [Vultr CCM](https://github.com/vultr/vultr-cloud-controller-manager)
  * Installs the [Vultr CSI](https://github.com/vultr/vultr-csi)

## Quick Start
Usage and input details can be found in the [Terraform Module Registry Docs](https://registry.terraform.io/modules/vultr/condor/vultr/latest), or use the quickstart below.

1) Create a `main.tf` file with the [Required Inputs](https://registry.terraform.io/modules/vultr/condor/vultr/latest?tab=inputs#required-inputs):
``` hcl
# main.tf

module "condor" {
  source                 = "vultr/condor/vultr"
  version                = "2.0.0"
  provisioner_public_key = chomp(file("~/.ssh/id_rsa.pub"))
  cluster_vultr_api_key  = var.cluster_vultr_api_key
  control_plane_firewall_rules = [
    {
      port    = 6443
      ip_type = "v4"
      source  = "0.0.0.0/0"
    }
  ]
}
```
  * The Control Plane Firewall rule in this example exposes the Kubernetes API globally, it is recommended you configure a more restrictive rule or rules on production clusters. 
  * Passing the Cluster API Key as plain text is not recommended for anything beyond testing, use an environment variable as described [here](https://www.terraform.io/docs/cli/config/environment-variables.html#tf_var_name).

2) Configure any [Optional Inputs](https://registry.terraform.io/modules/vultr/condor/vultr/latest?tab=inputs#optional-inputs) if you wish to change from the defaults.

3) Deploy
``` sh
$ terraform init && terraform apply
```

4) The Admin Kubeconfig is written locally to your working directory by default as `admin-<terraform-workspace>.conf`. This may not be desireable in some scenarios, e.g. running Terraform in automation(CI/CD pipelines). If you wish to change this set the `write_kubeconfig` variable to false. If you wish to retreive the kubeconfig manually, you can run the following after your cluster has been deployed:
``` sh
$ k0sctl kubeconfig > /path/to/admin.conf
```

5) Verify cluster functionality
``` sh
kubectl --kubeconfig admin.conf get no 
kubectl --kubeconfig admin.conf get po -n kube-system
```

## Firewall Configuration
### Control Plane HA VLB Firewall
The Control Plane LB Firewall is configured to allow only what is needed by the cluster as described in the [K0s Networking Docs](https://docs.k0sproject.io/v1.21.1+k0s.0/networking/#required-ports-and-protocols) by default. The Kubernetes API will not be accessible without configuring an additional rule or rules(as shown in the quickstart example) via the `control_plane_firewall_rules` input variable. E.g.:
``` hcl
  control_plane_firewall_rules = [
    {
      port    = 6443
      ip_type = "v4"
      source  = "0.0.0.0/0"
    }
  ]
```
As also stated in the quickstart, this example rule exposes the Kubernetes API globally, your rules should be more restrictive for production clusters.

### Cluster Nodes Vultr Firewall
The cluster nodes(control plane and workers) Vultr Firewall defaults to allowing only SSH globally. This is generally acceptable, however if you would like to restrict access further you may disable this rule by setting the `allow_ssh` input variable to `false` then configuring the desired rule/rules outside of this module using the `cluster_firewall_group_id` output in your rules. 

## Extensions
### Helm
Helm Repositories and Charts may be configured/deployed during initial cluster init via the `helm_repositories` and `helm_charts` variables. Note, namespace in a chart definition is required for K0s. Example:
``` hcl
  helm_repositories = [
    {
      name = "argo"
      url  = "https://argoproj.github.io/argo-helm"
    }
  ]
  helm_charts = [
    {
      name      = "argocd"
      chartname = "argo/argo-cd"
      version   = "3.6.8"
      namespace = "argocd"
    }
  ]
```
Please see the [Helm Chart Deployer](https://docs.k0sproject.io/v1.21.3+k0s.0/helm-charts/#helm-charts) docs for a comprehensive list of field/parameter values and further details. Note, this feature entails [Limitagions](#limitagions).

### Manifest Bundles
You may deploy any Kubernetes manifests automatically with the [K0s Manifest Deployer](https://docs.k0sproject.io/v1.21.1+k0s.0/manifests/#manifest-deployer) by placing your manifests in the `/var/lib/k0s/manifests` directory. Doing so via this module is not supported, however you may use the resulting `controller_ips` module output as arguments to a separate module that copies your manifests to the specified directory(or as stated in the linked K0s docs, a "stack" subdirectory).

## Limitations
* Shrinking of the Control Plane is not supported, only growing. You will need to manually run `k0s etcd leave` on all Control Plane nodes with index > 0 prior to shrinking the `controller_count`. An initial attempt was made to implement this in a destroy time provisioner, however it caused issues when running `terraform destroy` to destroy the entire plan. This may be revisited at a later date. 
* Etcd running on the Control Plane currently goes over the public network due to k0sctl configuration limitaitons - each controllers respective `spec.k0s.spec.storage.peerAddress` parameters would require different values. The Cluster Vultr Firewall allows only Control plane nodes to access the Etcd port(`2380`). Will likely revisit this or look into a PR to K0sctl so that the private network may be used.
* Helm Charts/Repositories are not removed from your cluster if removed from your Terraform configuration. The manifests must be manually removed from each controller in the `/var/lib/k0s/manifests/helm/` directory, and the resources/namespace manually deleted. This makes it less than ideal for continued deployments to your cluster, but great for bootstrapping core applications. A tool such as ArgoCD as used in the Helm example is recommended for long term deployments to your cluster. 


CHANGELOG.md
# Change Log
## [v2.0.0](https://github.com/vultr/terraform-vultr-condor/releases/tag/v2.0.0) (2021-09-03)
### Breaking Changes
* Condor is now based on Mirantis K0s rather than Kubeadm, as such v2 is completely incompatible with previous releases.
### Features
* HA Control Plane
* Isolated control plane(Control Plane nodes are not part of the cluster)
* Control Plane and Worker Node firewalls. 
* Declarative cluster and component(Vultr CCM, CSI, Calico, etc.) upgrades
* K0s manifest deployer support 

## [v1.3.0](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.3.0) (2021-08-25)
### Breaking Changes
* Remove `calico_wireguard` variable - nodes were not properly configured, will need to revisit
* Add `calico_mode` variable - previously defaulted to `vxlan`, is now configurable but defaults to `bird`
### Changes
* Bump K0s release from `v1.21.1+k0s.0` to `v1.21.3+k0s.0`
* Bump kube-system component versions

## [v1.2.3](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.2.3) (2021-07-14)
### Fixes
* Handle dynamic NIC names.

## [v1.2.2](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.2.2) (2021-06-21)
### Fixes
* Fix firewall configuration after Vultr image changes. 

## [v1.2.1](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.2.1) (2021-06-19)
### Changes
* Template Vultr CSI 
* Add Vultr CSI image/version vars
* Add kubeconfig filename tf workspace suffix

## [v1.2.0](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.2.0) (2021-06-12)
### Features
* Support K0s Helm deployments.
### Changes
* Convert module internal K0sctl configuration to HCL from YAML.
### Fixes
* Change Controller/Worker network interfaces from ens3/ens7 to enp1s0/enp6s0 due to Vultr image changes. 

## [v1.1.0](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.1.0) (2021-06-06)
### Features
* Write Kubeconfig locally option.
* Control Plane VLB Firewall. 
### Changes
* Add variable descriptions.
* Lock up cluster firewall, SSH only by default. 
* Docs updates.
### Fixes
* README markdown.

## [v1.0.1](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.0.1) (2021-06-05)
### Fixes
* Remove unused variables from triggers map.

## [v1.0.0](https://github.com/3letteragency/terraform-vultr-k0s/releases/tag/v1.0.0) (2021-06-05)
### First Release


scripts/provision.sh
#!/usr/bin/env bash
set -euxo posix

safe_apt(){
	while fuser /var/{lib/{dpkg,apt/lists},cache/apt/archives}/lock >/dev/null 2>&1 ; do
		echo "Waiting for apt lock..."
		sleep 1
	done
	apt "$@"
}

safe_apt -y update
safe_apt -y install jq

PUBLIC_MAC=$(curl --silent 169.254.169.254/v1.json | jq -r '.interfaces[0].mac')
PUBLIC_NIC=$(ip -j link | jq --arg PUBLIC_MAC $PUBLIC_MAC '.[] | select(.address==$PUBLIC_MAC)' | jq -r .ifname)
INTERNAL_MAC=$(curl --silent 169.254.169.254/v1.json | jq -r '.interfaces[1].mac')
INTERNAL_NIC=$(ip -j link | jq --arg INTERNAL_MAC $INTERNAL_MAC '.[] | select(.address==$INTERNAL_MAC)' | jq -r .ifname)
INTERNAL_IP=$1
CONTROL_PLANE_PORTS=(6443 2379 2380 10250 10251 10252 8132 8133 9443)

if [ $(echo $HOSTNAME | grep controller) ]; then
    NODE_ROLE=controller
elif [ $(echo $HOSTNAME | grep worker) ]; then
    NODE_ROLE=worker
fi

case $NODE_ROLE in
    controller)
        for port in "${CONTROL_PLANE_PORTS[@]}"; do
            ufw allow $port
            ufw allow in on $INTERNAL_NIC
        done
    ;;
    worker)
        ufw allow 10250
        ufw allow 179
        ufw allow 9443
        ufw allow 4789/udp
        ufw allow 8132:8133/tcp
        ufw allow 30000:32767/tcp
        ufw allow in on $INTERNAL_NIC
    ;;
esac

ufw reload

cat <<-EOF > /etc/systemd/network/public.network
  [Match]
  MACAddress=$PUBLIC_MAC

  [Network]
  DHCP=yes
EOF

cat <<-EOF > /etc/systemd/network/private.network
  [Match]
  MACAddress=$INTERNAL_MAC

  [Network]
  Address=$INTERNAL_IP
EOF

echo "# For k0s"                            >> /etc/hosts
echo "$INTERNAL_IP             $(hostname)" >> /etc/hosts

systemctl enable systemd-networkd systemd-resolved
systemctl restart systemd-networkd systemd-resolved
systemctl disable networking


scripts/common/remote/common-provisioner.sh
#!/bin/bash

set -euxo posix

if [ $# -ne 3 ]; then
	echo "common-provisioner.sh requires 3 parameters"
	exit 1
fi

yum -y update
yum -y install epel-release
yum -y install jq curl

INSTANCE_METADATA=$(curl --silent http://169.254.169.254/v1.json)
PRIVATE_IP=$(echo $INSTANCE_METADATA | jq -r .interfaces[1].ipv4.address)
PUBLIC_MAC=$(curl --silent 169.254.169.254/v1.json | jq -r '.interfaces[] | select(.["network-type"]=="public") | .mac')
PRIVATE_MAC=$(curl --silent 169.254.169.254/v1.json | jq -r '.interfaces[] | select(.["network-type"]=="private") | .mac')

# Parameters
DOCKER_RELEASE="$1"
CONTAINERD_RELEASE="$2"
K8_RELEASE=$(echo $3 | sed 's/v//')

pre_dependencies(){
	cat <<-EOF > /etc/sysctl.d/k8s.conf
		net.bridge.bridge-nf-call-ip6tables = 1
		net.bridge.bridge-nf-call-iptables = 1
		EOF

	sysctl --system

	systemctl disable firewalld
	systemctl stop firewalld
}

network_config(){
	yum -y install systemd-networkd systemd-resolved
	systemctl disable network NetworkManager
	rm -f /etc/resolv.conf
	ln -s /run/systemd/resolve/resolv.conf /etc/resolv.conf
	mkdir /etc/systemd/network

	cat <<-EOF > /etc/systemd/network/public.network
		[Match]
		MACAddress=$PUBLIC_MAC

		[Network]
		DHCP=yes
		EOF

	cat <<-EOF > /etc/systemd/network/private.network
		[Match]
		MACAddress=$PRIVATE_MAC

		[Network]
		Address=$PRIVATE_IP
		EOF

	systemctl enable systemd-networkd systemd-resolved
	systemctl restart systemd-networkd systemd-resolved
}

install_k8(){
	cat <<-EOF > /etc/yum.repos.d/kubernetes.repo
		[kubernetes]
		name=Kubernetes
		baseurl=https://packages.cloud.google.com/yum/repos/kubernetes-el7-x86_64
		enabled=1
		gpgcheck=1
		repo_gpgcheck=1
		gpgkey=https://packages.cloud.google.com/yum/doc/yum-key.gpg https://packages.cloud.google.com/yum/doc/rpm-package-key.gpg
		EOF

	cat <<-EOF > /etc/sysconfig/kubelet
		KUBELET_EXTRA_ARGS="--cloud-provider=external"
		EOF

	# setenforce 0
	sed -i 's/^SELINUX=enforcing$/SELINUX=permissive/' /etc/selinux/config

	yum install -y kubelet-$K8_RELEASE kubeadm-$K8_RELEASE kubectl-$K8_RELEASE --disableexcludes=kubernetes

	systemctl daemon-reload
	systemctl enable kubelet
}

install_docker(){
	yum install -y yum-utils device-mapper-persistent-data
	yum-config-manager --add-repo https://download.docker.com/linux/centos/docker-ce.repo
	yum -y install containerd.io-$CONTAINERD_RELEASE docker-ce-$DOCKER_RELEASE docker-ce-$DOCKER_RELEASE

	if test -d /etc/docker; then echo "continuing"; else mkdir /etc/docker; fi

	cat <<-EOF > /etc/docker/daemon.json
		{
		  "exec-opts": ["native.cgroupdriver=systemd"],
		  "log-driver": "json-file",
		  "log-opts": {
		    "max-size": "100m"
		  },
		  "storage-driver": "overlay2"
		}
		EOF

	mkdir -p /etc/systemd/system/docker.service.d

	systemctl daemon-reload
	systemctl enable docker
	systemctl restart docker
}

clean(){
	rm -f /tmp/common-provisioner.sh
}

main(){
	pre_dependencies

	network_config
	install_docker
	install_k8
	clean
}

main



scripts/condor-provision.sh
#!/usr/bin/env bash
set -euxo posix

safe_apt(){
	while fuser /var/{lib/{dpkg,apt/lists},cache/apt/archives}/lock >/dev/null 2>&1 ; do
		echo "Waiting for apt lock..."
		sleep 1
	done
	apt "$@"
}

safe_apt -y update
safe_apt -y install jq gnupg2

INSTANCE_METADATA=$(curl --silent http://169.254.169.254/v1.json)
PRIVATE_IP=$(echo $INSTANCE_METADATA | jq -r .interfaces[1].ipv4.address)
PUBLIC_MAC=$(curl --silent 169.254.169.254/v1.json | jq -r '.interfaces[] | select(.["network-type"]=="public") | .mac')
PRIVATE_MAC=$(curl --silent 169.254.169.254/v1.json | jq -r '.interfaces[] | select(.["network-type"]=="private") | .mac')
HOSTNAME=$(curl --silent 169.254.169.254/v1.json | jq -r '.hostname')

if [ $(echo $HOSTNAME | grep controller) ]; then
	NODE_ROLE=controller
elif [ $(echo $HOSTNAME | grep worker) ]; then
	NODE_ROLE=worker
fi

CONTAINERD_RELEASE="${CONTAINERD_RELEASE}"
K8_VERSION="${K8_VERSION}"
PRE_PROVISIONED=${PRE_PROVISIONED}
FILES_TO_CLEAN="/tmp/condor-provision.sh"
CONTROL_PLANE_PORTS=(6443 2379 2380 10250 10251 10252)

set_hostname(){
	echo $HOSTNAME > /etc/hostname
	hostnamectl set-hostname $HOSTNAME
}

system_config(){
	cat <<-EOF > /etc/modules-load.d/containerd.conf
		overlay
		br_netfilter
		EOF

	cat <<-EOF > /etc/sysctl.d/99-kubernetes-cri.conf
		net.bridge.bridge-nf-call-iptables  = 1
		net.ipv4.ip_forward                 = 1
		net.bridge.bridge-nf-call-ip6tables = 1
		EOF

	modprobe overlay
	modprobe br_netfilter
	sysctl --system
}

firewall_config(){
	case $NODE_ROLE in
		controller)
			for port in "$${CONTROL_PLANE_PORTS[@]}"; do
				ufw allow $port
				ufw allow in on ens7
			done
			;;
		worker)
			ufw allow 10250
			ufw allow 30000:32767/tcp
			ufw allow in on ens7
			;;
	esac

	ufw reload
}

network_config(){
	firewall_config

	cat <<-EOF > /etc/systemd/network/public.network
		[Match]
		Name=ens3

		[Network]
		DHCP=yes
		EOF

	cat <<-EOF > /etc/systemd/network/private.network
		[Match]
		Name=ens7

		[Network]
		Address=$PRIVATE_IP
		EOF

	systemctl enable systemd-networkd systemd-resolved
	systemctl restart systemd-networkd systemd-resolved
	systemctl disable networking
}

install_containerd(){
	safe_apt -y update
	safe_apt -y install apt-transport-https ca-certificates curl software-properties-common

	curl -fsSL https://download.docker.com/linux/debian/gpg | sudo apt-key --keyring /etc/apt/trusted.gpg.d/docker.gpg add -

    add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable"

	safe_apt -y update
	safe_apt -y install containerd.io=$CONTAINERD_RELEASE
}

install_k8(){
	curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -

	cat <<-EOF > /etc/apt/sources.list.d/kubernetes.list
		deb https://apt.kubernetes.io/ kubernetes-xenial main
		EOF

	safe_apt -y update
	safe_apt -y install kubelet=$K8_VERSION kubeadm=$K8_VERSION kubectl=$K8_VERSION
	apt-mark hold kubelet kubeadm kubectl

	cat <<-EOF > /etc/default/kubelet
		KUBELET_EXTRA_ARGS="--cloud-provider=external"
		EOF
}

clean(){
	rm -f $FILES_TO_CLEAN
}

main(){
	if [ "$PRE_PROVISIONED" = true ]; then
	    set_hostname
	    network_config
	else
		system_config
		network_config
		install_containerd
		install_k8
	fi

	clean
}

main
